# SECOND HOMEWORK: NETWORK OPTIMIZATION AND NON-LINEAR MODELS (TOPICS 2 AND 3)

## Data Science and Engineering



**Full name:** Laura Escobar Crespo
    

---

### NETWORK OPTIMIZATION PROBLEM

---

**Description of the problem:**

This problem simulates a logistic problem that sell a given and unknown product X. In this case, we have six customers, two factories and four stores. Each customer has a specific demand of units of the product. This can be satisfied from one or more of the four stores, or directly from one of both two factories (or even a combination of stores and factories). 

Factories are the ones which produces the goods; however, they have a maximum amount of production. Similarly, each store can support a maximum volume of product moving through it.

Each product that arraives to a store should pass a quality control. However each laboratoy only has capacity for 10,000 units, and the number of laboratories in each store varies. 

Products can go from a factory to a client, from a factory to a store or from a store to a client. Each of these relations of transport have a different cost per unit, and depends which factory, store or client are we dealing with.

The factories are located in Madrid and Barcelona, whereas the store are in Sevilla, Toledo, Valencia and Zaragoza. For privacy reasons, the six clients are just enumerated (C1, C2, C3, C4, C5 and C6).

The capacity (named *supplies*) of the factories is given by the following table:

| Factory | Supplies (untis) |
| --- | --- |
| Madrid | 150,000 |
| Barcelona |  200,000 |


As it was described before, the product can be shipped from a factory to a set of four stores. Each store has a maximum storage capacity. Stores do not produce or demand the product; they simply pass the product to customers. So any amount of product given to a customer from a store should be equal to the given from a factory to that store.

| Store | Capacity (units) |
| --- | --- |
| Sevilla | 70,000 |
| Toledo | 50,000 |
| Valencia | 100,000 |
| Zaragoza | 40,000 |

Now, each store has a specific number of laboratories in each store:

| Store | Laboratories |
| --- | --- |
| Sevilla | 5 |
| Toledo | 6 |
| Valencia | 5 |
| Zaragoza | 5 |

And, finally, the customers' demand:

| Customer | Demand (units) |
| --- | --- |
| C1 | 50,000 |
| C2 | 10,000 |
| C3 | 40,000 |
| C4 | 35,000 |
| C5 | 60,000 |
| C6 | 20,000 |

The following table shows the cost, in euros per unit, of the relation between the starting point (factory or store) and the end point (store or client). Columns are the starting points and rows are the end points. Some cells are filled with '-', meaning that the combination is not possible. The emty cells are considered as ceros, or useless relations. They are the relation between two stores, which is not going to be in the optimal solution as it will only increase the costs, but they are neccesary to be defined in order our model work.

| To | Madrid | Barcelona | Sevilla | Toledo | Valencia | Zaragoza |
| --- | --- | --- | --- | --- | --- | --- |
| Sevilla  | 0.5 |   - |
| Toledo | 0.5 | 0.3 |
| Valencia     | 1.0 | 0.5 |
| Zaragoza     | 0.2 | 0.2 |
| C1 | 1.0 | 2.0 |   - | 1.0 |   - |   - |
| C2 |   - |   - | 1.5 | 0.5 | 1.5 |   - |
| C3 | 1.5 |   - | 0.5 | 0.5 | 2.0 | 0.2 |
| C4 | 2.0 |   - |   - | 1.0 |   - | 1.5 |
| C5 |   - |   - | 1.5 | 0.5 | 0.5 | 0.5 |
| C6 | 1.0 |   - | 1.0 |   - | 1.5 | 1.5 |

The objective of this problem is to satify the demand of the clients but spending as less as possible (minimize the costs).

---

**Formulation of the problem:**



*Sets:*

$$f \in \text{Factories}=\{\text{Madrid}, \text{Barcelona}\}$$

$$s \in \text{Stores}=\{\text{Sevilla}, \text{Toledo}, \text{Valencia}, \text{Zaragoza}\}$$

$$c \in \text{Customers}=\{\text{C1}, \text{C2}, \text{C3}, \text{C4}, \text{C5}, \text{C6}\}$$

$$i \in \text{Origins} = \text{Factories} \cup \text{Stores}=\{\text{Madrid}, \text{Barcelona}, \text{Sevilla}, \text{Toledo}, \text{Valencia}, \text{Zaragoza}\}$$

$$j \in \text{Destinations} = \text{Stores} \cup \text{Customers}=\{\text{Sevilla}, \text{Toledo}, \text{Valencia}, \text{Zaragoza}, \text{C1}, \text{C2}, \text{C3}, \text{C4}, \text{C5}, \text{C6}\}$$

*Parameters:*

*   Maximum amount of units in a factory: $\hspace{0.2cm}\text{supplies}_f \in \mathbb{R}^+$


*   Demand of products by customers: $\hspace{0.2cm}\text{demand}_c \in \mathbb{R}^+$


*   Capacity of stores: $\hspace{0.2cm}\text{capacity}_s \in \mathbb{R}^+$


*   Laboratories of stores: $\hspace{0.2cm}\text{laboratory}_s \in \mathbb{R}^+$


*   Cost euros per unit of the relation between origins and destinations: $\hspace{0.2cm}\text{cost}_{i,j} \in \mathbb{R}^+$


*Decisional variables:*

Units of products that are sent from an origin $i$ to a destination $j$:

$$\text{flow}_{i,j} \in \mathbb{N}^+$$


*Objective function:*

The objective of this problem is to minimize the cost.

\begin{equation}
\text{min} \sum_{(i,j)}{\text{cost}_{i,j}·\text{flow}_{i,j}}
\end{equation}


*Constraints:*

*   Factory Supplies Constraint: The amount of products that are produce in a factory must respect the capacity of the factories.


\begin{equation}
\sum_{j \in \text{Destinations}}{\text{flow}_{f,j}} \leq \text{supplies}_{f} \quad \forall f \in \text{Factories}
\end{equation}


*   Customer Demand Constraint: The demand of each client must be satisfied.

\begin{equation}
\sum_{i \in \text{Origins}}{\text{flow}_{i,c}} = \text{demand}_{c} \quad \forall c \in \text{Customers}
\end{equation}


*   Store Capacity Constraint: The amount of products that enters in a store must respect its capacity.

\begin{equation}
\sum_{j \in \text{Destinations}}{\text{flow}_{s,j}} \leq \text{capacity}_{s}
\quad \forall s \in \text{Stores}
\end{equation}


*   Depot Flow Constraint: The number of products that come out from the store must be equal to the number of products that enters in.

\begin{equation}
\sum_{j \in \text{Destinations}}{\text{flow}_{s,j}} = 
\sum_{f \in \text{Factories}}{\text{flow}_{f,s}}
\quad \forall s \in \text{Stores}
\end{equation}

*   Laboratory Constrain: The number of laboratories times its capacity should not be greater that the number of products received.

\begin{equation}
\sum_{j \in \text{Destinations}}{\text{flow}_{s,j}} \leq \text{laboratories}_{s}·10,000
\quad \forall s \in \text{Stores}
\end{equation}



*Note 1:* As we are working always with units of products, we multiply the number of laboratories times their capacity (which is 10,000 units) to work with the same notation.

---

**Explanation of the code:**

After loading the Pyomo Modeling Enviroment, the *model* is defined. In this case the model is *Concrete* as it is inmediatily constructed and the used data	is	presented	at	the	time as the	components are	defined. It should be pointed out that the model is constructed using a local variable so it is easier to call when it is needed.

Next, the sets ranges are defined. As the given data is not too large (it cannot be too large because it is a network optimization problem and the computational time could spend too much time), are written manually.

Thanks to the sets, now we are able to built the parameters, which contains crucial information to solve the problem.

The decison varibles, named as *flow*, uses the index sets previusly created of the origins and destinations with domanin *NonNegativeReals*, which means that in the solution we do not accept negative numbers. As we know the the sets for these varibles, we know that at the end we are computing for 60 decision variables (origins multiplied by destinations is equal to 60).

*Note 2*: When the values for the cost are defined, some relations between an origin and a destination has a really big value. This is because, as it was explainded before, there are some combinations which are impossible, so in order to ensure that this impossible ways are not chosen, we stablish its cost with a ridiculous value.

The next step is to set the objective function. It is just following the formula defined above applied to python. It is important to specify our desire of minimize. The same is done for the five constraints, it is just following the formulas.

*Note 3:* To see in more detail, the definitions of the sets, parameters, decision variables, objective function and constraints, they are described in *Formulation of the problem*.

---

**Code:**

In [ ]:
from pyomo.environ import *
model = ConcreteModel()

model.i = Set(initialize=['Madrid','Barcelona','Sevilla','Toledo','Valencia','Zaragoza'], doc='origins' )
model.j = Set(initialize=['Sevilla','Toledo','Valencia','Zaragoza','C1', 'C2', 'C3','C4','C5', 'C6'], doc='destinations')

model.factory = Set(initialize=['Madrid','Barcelona'])
model.store = Set(initialize=['Sevilla','Toledo','Valencia','Zaragoza'])
model.customer = Set(initialize=['C1', 'C2', 'C3','C4','C5', 'C6'])

model.supplies = Param(model.factory, initialize={ 'Madrid':150000,'Barcelona':200000}) 
model.capacity = Param(model.store, initialize={'Sevilla':70000,'Toledo':50000,'Valencia':100000,'Zaragoza':40000})
model.laboraty = Param(model.store, initialize={'Sevilla':5,'Toledo':6,'Valencia':5,'Zaragoza':5})
model.demand = Param(model.customer, initialize={'C1':50000,'C2':10000,'C3': 40000,'C4' :35000,'C5':60000,'C6':20000})

cost = ({
    ('Madrid', 'C2'): 10000000,
    ('Madrid', 'Sevilla'): 0.5,
    ('Madrid', 'Toledo'): 0.5,
    ('Madrid', 'Valencia'): 1.0,
    ('Madrid', 'Zaragoza'): 0.2,
    ('Madrid', 'C1'): 1.0,
    ('Madrid', 'C3'): 1.5,
    ('Madrid', 'C4'): 2.0,
    ('Madrid', 'C6'): 1.0,
    ('Madrid', 'C5'): 10000000,
    ('Barcelona', 'Toledo'): 0.3,
    ('Barcelona', 'Sevilla'): 10000000,
    ('Barcelona', 'Valencia'): 0.5,
    ('Barcelona', 'Zaragoza'): 0.2,
    ('Barcelona', 'C1'): 2.0,
    ('Barcelona', 'C2'): 10000000,
    ('Barcelona', 'C3'): 10000000,
    ('Barcelona', 'C4'): 10000000,
    ('Barcelona', 'C5'): 10000000,
    ('Barcelona', 'C6'): 10000000,
    ('Sevilla', 'Valencia'): 0,
    ('Sevilla', 'Sevilla'): 0,
    ('Sevilla', 'Zaragoza'): 0,
    ('Sevilla', 'Toledo'): 0,
    ('Sevilla', 'C4'): 10000000,
    ('Sevilla', 'C1'): 10000000,
    ('Sevilla', 'C2'): 1.5,
    ('Sevilla', 'C3'): 0.5,
    ('Sevilla', 'C5'): 1.5,
    ('Sevilla', 'C6'): 1.0,
    ('Toledo', 'Valencia'): 0,
    ('Toledo', 'Zaragoza'): 0,
    ('Toledo', 'Sevilla'): 0,
    ('Toledo', 'Toledo'): 0,
    ('Toledo', 'C1'): 1.0,
    ('Toledo', 'C2'): 0.5,
    ('Toledo', 'C3'): 0.5,
    ('Toledo', 'C4'): 1.0,
    ('Toledo', 'C5'): 0.5,
    ('Toledo', 'C6'): 10000000,
    ('Valencia', 'Valencia'): 0,
    ('Valencia', 'Zaragoza'): 0,
    ('Valencia', 'Sevilla'): 0,
    ('Valencia', 'Toledo'): 0,
    ('Valencia', 'C1'): 10000000,
    ('Valencia', 'C4'): 10000000,
    ('Valencia', 'C2'): 1.5,
    ('Valencia', 'C3'): 2.0,
    ('Valencia', 'C5'): 0.5,
    ('Valencia', 'C6'): 1.5,
    ('Zaragoza', 'Valencia'): 0,
    ('Zaragoza', 'Sevilla'): 0,
    ('Zaragoza', 'Zaragoza'): 0,
    ('Zaragoza', 'Toledo'): 0,
    ('Zaragoza', 'C1'): 10000000,
    ('Zaragoza', 'C2'): 10000000,
    ('Zaragoza', 'C3'): 0.2,
    ('Zaragoza', 'C4'): 1.5,
    ('Zaragoza', 'C5'): 0.5,
    ('Zaragoza', 'C6'): 1.5
})

model.cost = Param(model.i, model.j, initialize=cost)
model.flow = Var(model.i, model.j, within=NonNegativeReals)

In [ ]:
def objective_rule(model):
  return sum(model.cost[i,j]*model.flow[i,j] for i in model.i for j in model.j)

model.objective = Objective(rule=objective_rule, sense=minimize)

In [ ]:
def FactoryConstraint(model, f):
  return sum(model.flow[f,j] for j in model.j) <= model.supplies[f]

model.Factory = Constraint(model.factory, rule=FactoryConstraint)
    
def DemandConstraint(model, c):
  return sum(model.flow[i,c] for i in model.i) == model.demand[c]

model.Demand = Constraint(model.customer, rule=DemandConstraint)


def StoreCapacityConstraint(model, s):
  return sum(model.flow[s,j] for j in model.j)<=model.capacity[s]

model.StoreCapacity = Constraint(model.store, rule=StoreCapacityConstraint)


def StoreConstraint(model, s):
  return (sum(model.flow[s,j] for j in model.j)==sum(model.flow[f,s] for f in model.factory))

model.Store = Constraint(model.store, rule=StoreConstraint)


def LaboratyConstraint(model, s):
  return sum(model.flow[s,j] for j in model.j)<=model.laboraty[s]*10000

model.Laboraty = Constraint(model.store, rule=LaboratyConstraint)



---

**Show results:**

In [ ]:
Solver = SolverFactory('glpk')
Results = Solver.solve(model)

# Display solution
model.flow.display()

flow : Size=60, Index=flow_index
    Key                       : Lower : Value   : Upper : Fixed : Stale : Domain
          ('Barcelona', 'C1') :     0 :     0.0 :  None : False : False : NonNegativeReals
          ('Barcelona', 'C2') :     0 :     0.0 :  None : False : False : NonNegativeReals
          ('Barcelona', 'C3') :     0 :     0.0 :  None : False : False : NonNegativeReals
          ('Barcelona', 'C4') :     0 :     0.0 :  None : False : False : NonNegativeReals
          ('Barcelona', 'C5') :     0 :     0.0 :  None : False : False : NonNegativeReals
          ('Barcelona', 'C6') :     0 :     0.0 :  None : False : False : NonNegativeReals
     ('Barcelona', 'Sevilla') :     0 :     0.0 :  None : False : False : NonNegativeReals
      ('Barcelona', 'Toledo') :     0 : 50000.0 :  None : False : False : NonNegativeReals
    ('Barcelona', 'Valencia') :     0 : 50000.0 :  None : False : False : NonNegativeReals
    ('Barcelona', 'Zaragoza') :     0 : 40000.0 :  None : False : F

In [ ]:
print('Total costs:', model.objective())

Total costs: 200000.0


---

**Explanation of the results:**

For this problem it is used the *glpk* solver.

As it was expected a lot of decision variables are zero:

| To | Madrid | Barcelona | Sevilla | Toledo | Valencia | Zaragoza |
| --- | --- | --- | --- | --- | --- | --- |
| Sevilla  |  5,000 |   - |
| Toledo |  | 50,000 |
| Valencia     |  | 50,000 |
| Zaragoza     |  | 40,000 |
| C1 | 50,000 |  |   - |  |   - |   - |
| C2 |   - |   - |  | 10,000 |  |   - |
| C3 |  |   - | 5,000 |  |  | 35,000 |
| C4 |  |   - |   - | 35,000 |   - |  |
| C5 |   - |   - |  | 5,000 | 50,000 | 5,000 |
| C6 | 20,000 |   - |  |   - |  |  |


If we check all the values, we see that the customers' demand is satisfied. Also the capacity for each factory, laboratory and store is never surpass, meaning that the constraints works as they stablish limits. Also, defining a big value for the cost works great, ensuring that the value for these decision variables are cero (they are the impossible ways). Finally, it is also satisfied the flow in the stores (the number of products that come out from the store must be equal to the number of products that enters in), as for example, the store in Sevilla sends 5,00 units of the product to Customer 2, but it recives the 5,000 units from the factory of Madrid.

---

### NON-LINEAR MODEL

---

**Description of the problem:**

A company in juices market produces five different types of juices (A, B, C, D, E and F). Each juice has a specific quatity of pure juice, aditives and water in order to make each one distintive (some of them are obtained with less quatity of total liters, which is the sum of pure juice, water and aditives). Besides, each juice has a production cost.

The information about the juices is given by:

| Juice | Pure juice (litres) | Total litres | Cost (euros per litre) |
| --- | --- | --- | --- |
| A | 2 | 4 | 10 |
| B | 5 | 9 | 2 |
| C | 3 | 6 | 3 |
| D | 1 | 3 | 2 |
| E | 0.5 | 2 | 1 |
| F | 1.5 | 5 | 1 |


It is also known that the company has only 20,000 kg of fruit. Knowing that from 10 kg of fruit they obtain 7 litres of pure juice (then 20,000x7/10 = 14,000 litres of pure juice).

It is know that for marketing reasons that at least a 20% of the total litres should be of juice A and 40% should be of juice B. The rest have to reach the 3,000 total litres each.

The benefits for the company is given by the equation:

\begin{equation}
2· ({\text{a}^2 + \text{b}^2 + \text{c}^2 +\text{d}^2 +\text{e}^2 +\text{f}^2 })
\end{equation}

where a, b, c, d, e and f are the total final litres of each type of juice.

The company wants to know how much of each type of juice has to produce in order to obtain the highest profit.

*Note 4:* In order to calculate the real benefit for the company, it must be taken into account the production costs.



---

**Formulation of the problem:** 

*Sets:*

$$o \in \text{Juice} = \{\text{A}, \text{B}, \text{C}, \text{D}, \text{E}, \text{F}\}$$
$$j \in \text{Destinations} = \text{Stores} \cup \text{Customers}=\{\text{Sevilla}, \text{Toledo}, \text{Valencia}, \text{Zaragoza}, \text{C1}, \text{C2}, \text{C3}, \text{C4}, \text{C5}, \text{C6}\}$$

*Parameters:*

*   Litres of pure juice: $\hspace{0.2cm}\text{pure juice}_o \in \mathbb{R}^+$


*   Total litres to create each type of juice: $\hspace{0.2cm}\text{litres}_o \in \mathbb{R}^+$


*   Cost of production of each type of juice: $\hspace{0.2cm}\text{cost}_s \in \mathbb{R}^+$

*Note 5:* Six more parameters are created (named a, b, c, d, e, f). They just contain one 1 and the rest are 0, because they are just telling the type of juice (1 is true and 0 is false). For example for parameter a, there is just a 1 in the first position and the rest are zeros.

*   $\text{a}_o$

*   $\text{b}_o$

*   $\text{c}_o$

*   $\text{d}_o$

*   $\text{e}_o$

*   $\text{f}_o$


*Decisional variables:*

Total litres for type of juice $o$:


$$\text{x}_{o} \in \mathbb{R}^+$$


*Objective function:*

The objective of this problem is to maximize the profit.

\begin{equation}
\text{max} \sum_{o \in \text{juices}}2·{\text{x}_{o}^2}-\sum_{o \in \text{juices}}{\text{x}_{o}}·{\text{cost}_{o}}
\end{equation}


*Constraints:*

*   Constraint for the pure juice: We have a limited volume of pure juice, given by fruit. The litres of pure juice used in the production cannot be higher than the stock.

\begin{equation}
\sum_{o \in \text{juices}} \frac{\text{pure juice}_{o}}{\text{litres}_{o}}· {\text{x}_{o}}\geq 14,000
\end{equation}


*   Constraint for juice A: Its final total litres should be at least a 20% of the total litres of juice produced.


\begin{equation}
\sum_{o \in \text{juices}}{\text{x}_{o}}·{\text{a}_{o}} \geq  \sum_{o \in \text{juices}}{\text{x}_{o}}·{\text{juices}_{o}}·0.2
\end{equation}


*   Constraint for juice B: Its final total litres should be at least a 40% of the total litres of juice produced.

\begin{equation}
\sum_{o \in \text{juices}}{\text{x}_{o}}·{\text{b}_{o}} \geq  \sum_{o \in \text{juices}}{\text{x}_{o}}·{\text{juices}_{o}}·0.4
\end{equation}



*   Constraint for juice C: Its total litres shoul be at least 3,000 litres

\begin{equation}
\sum_{o \in \text{juices}}{\text{x}_{o}}·{\text{c}_{o}} \geq  3,000
\end{equation}


*   Constraint for juice D: Its total litres shoul be at least 3,000 litres

\begin{equation}
\sum_{o \in \text{juices}}{\text{x}_{o}}·{\text{d}_{o}} \geq  3,000
\end{equation}

*   Constraint for juice E: Its total litres shoul be at least 3,000 litres

\begin{equation}
\sum_{o \in \text{juices}}{\text{x}_{o}}·{\text{e}_{o}} \geq  3,000
\end{equation}

*   Constraint for juice F: Its total litres shoul be at least 3,000 litres

\begin{equation}
\sum_{o \in \text{juices}}{\text{x}_{o}}·{\text{f}_{o}} \geq  3,000
\end{equation}


*Note 5: Constraints A and B follow the same structure. Constraints C, D, E and F are also very similar between each other. 

---

**Explanation of the code:**

After loading the Pyomo Modeling Enviroment, the *model* is defined. In this case the model is *Concrete* as it is inmediatily constructed and the used data	is	presented	at	the	time as the	components are	defined. It should be pointed out that the model is constructed using a local variable so it is easier to call when it is needed.

Next, the set range is defined. As the given data is not too large, is written manually.

Thanks to the set, now we are able to built the parameters, which contains crucial information to solve the problem. As it was said before, six parameters just contain one 1 and the rest are 0, because they are just telling the type of juice (1 is true and 0 is false).

The decison varibles, named as *x*, uses the index set previusly created, with domanin *NonNegativeReals*, which means that in the solution we do not accept negative numbers. As we know the the set for these varibles, we know that at the end we are computing for 5 decision variables (as much as types of juice we have).

The next step is to set the objective function. It is just following the formula defined above applied to python. It is important to specify our desire of maximize. The same is done for the seven constraints, it is just following the formulas.

*Note 6:* To see in more detail, the definitions of the sets, parameters, decision variables, objective function and constraints, they are described in *Formulation of the problem*.



---

**Code:**

In [ ]:
from pyomo.environ import *
model_nonlinear = ConcreteModel()



model_nonlinear.juice = Set(initialize=['A','B','C','D','E','F'])
model_nonlinear.purejuice = Param(model_nonlinear.juice, initialize={'A':2,'B':5,'C':3,'D':1,'E':0.5,'F':1.5}) 
model_nonlinear.litres = Param(model_nonlinear.juice, initialize={'A':4,'B':9,'C':6,'D':3,'E':2,'F':5})  
model_nonlinear.cost = Param(model_nonlinear.juice,initialize={'A':10,'B':2,'C':3,'D':2,'E':1,'F':1})
model_nonlinear.a = Param(model_nonlinear.juice,initialize={'A':1,'B':0,'C':0,'D':0,'E':0,'F':0})
model_nonlinear.b = Param(model_nonlinear.juice,initialize={'A':0,'B':1,'C':0,'D':0,'E':0,'F':0})
model_nonlinear.c = Param(model_nonlinear.juice,initialize={'A':0,'B':0,'C':1,'D':0,'E':0,'F':0})
model_nonlinear.d = Param(model_nonlinear.juice,initialize={'A':0,'B':0,'C':0,'D':1,'E':0,'F':0})
model_nonlinear.e = Param(model_nonlinear.juice,initialize={'A':0,'B':0,'C':0,'D':0,'E':1,'F':0})
model_nonlinear.f = Param(model_nonlinear.juice,initialize={'A':0,'B':0,'C':0,'D':0,'E':0,'F':1})


model_nonlinear.x = Var(model_nonlinear.juice, within=NonNegativeReals)

In [ ]:
def obj(model_nonlinear):
  return sum(2*(model_nonlinear.x[o]**2)-(model_nonlinear.x[o]*model_nonlinear.cost[o]) for o in model_nonlinear.juice)

model_nonlinear.Obj = Objective(rule=obj, sense=maximize)

In [ ]:
def Total_kg_fruit_Constraint(model_nonlinear):
  return sum((model_nonlinear.purejuice[o]/model_nonlinear.litres[o])*model_nonlinear.x[o] for o in model_nonlinear.juice)<=14000

model_nonlinear.Total_kg_fruit = Constraint(rule=Total_kg_fruit_Constraint)

def constraintC(model_nonlinear):
    return (sum(model_nonlinear.x[o]*model_nonlinear.c[o] for o in model_nonlinear.juice)) >= 3000

model_nonlinear.ConstraintC = Constraint(rule=constraintC)

def constraintD(model_nonlinear):
    return (sum(model_nonlinear.x[o]*model_nonlinear.d[o] for o in model_nonlinear.juice)) >= 3000

model_nonlinear.ConstraintD = Constraint(rule=constraintD)

def constraintE(model_nonlinear):
    return (sum(model_nonlinear.x[o]*model_nonlinear.e[o] for o in model_nonlinear.juice)) >= 3000

model_nonlinear.ConstraintE = Constraint(rule=constraintE)

def constraintF(model_nonlinear):
    return (sum(model_nonlinear.x[o]*model_nonlinear.f[o] for o in model_nonlinear.juice)) >= 3000

model_nonlinear.ConstraintF = Constraint(rule=constraintF)

def constraintA(model_nonlinear):
    return (sum(model_nonlinear.x[o]*model_nonlinear.a[o] for o in model_nonlinear.juice)) >= (sum(model_nonlinear.x[o] for o in model_nonlinear.juice))*0.2

model_nonlinear.ConstraintA = Constraint(rule=constraintA)

def constraintB(model_nonlinear):
    return (sum(model_nonlinear.x[o]*model_nonlinear.b[o] for o in model_nonlinear.juice)) >= 0.4*(sum(model_nonlinear.x[o] for o in model_nonlinear.juice))

model_nonlinear.ConstraintB = Constraint(rule=constraintB)


---

**Show resuts:**

In [ ]:
from pyomo.opt import SolverFactory
solver = SolverFactory('ipopt')
results = solver.solve(model_nonlinear)
model_nonlinear.x.display()

x : Size=6, Index=juice
    Key : Lower : Value              : Upper : Fixed : Stale : Domain
      A :     0 :   6067.34698094898 :  None : False : False : NonNegativeReals
      B :     0 : 12269.388043845918 :  None : False : False : NonNegativeReals
      C :     0 :         2999.99997 :  None : False : False : NonNegativeReals
      D :     0 :         2999.99997 :  None : False : False : NonNegativeReals
      E :     0 :         2999.99997 :  None : False : False : NonNegativeReals
      F :     0 :         2999.99997 :  None : False : False : NonNegativeReals


In [ ]:
print('Benefits: ', model_nonlinear.Obj())

Benefits:  446594951.0297125


---

**Explanation of the results:**

If we check all the total litres for each type of juice, we see that the constraints are satisfied and it achieves the greatest profit possible. For example, juice C, D, E and F are exaclty 3,000 litres each. Also, (I calculated this with the calculator) the 20% of the total litres of the production is about 5,467 which is less than the total litres of juice A, so Constraint A is also satisfied. 

The results (the values are rounded):

| Juice | Total litres | 
| --- | --- | 
| A | 6,067 | 
| B | 12,269 | 
| C | 3,000 | 
| D | 3,000 | 
| E | 3,000 | 
| F | 3,000 | 

---

### CONCLUSIONS ABOUT THE SECOND HOMEWORK

---

This assignment is about half of the topic 2 and the entire content of topic 3.

As the First Homework, the hardest part of the assignment is to come up with an interesting problem for each case. 

Regarding to the first part of the work, the Network problem, there are few different type of exercises. Probably, the most well-known are those related with transport and logistics. In my case, I have chosed this type of problem, but I have also based it in the *Minimum Cost Flow Problem*. In fact, its base is practically the same as the algorithm searches for the combination of decision variables that gives the minimum cost. The differnt part is the definition of the constraints (and the needed sets and parameters) that add an extra difficulty to the mentioned problem. Although the data is not real (I invented the values), the result and conclusion of the problem is quite realistic if we look to a real case. The products are sent to the clients with the journey with less, we try to not overpass the capacities of the factories and stores and the customer's demand is satisfied.


On the other hand, defining a Non-lineal Model has been a real challenge for me. Although, and making some constructive critisim, I would loved to prepare a more complex problem (not in size data, I am talking about more complex constraints), I have to addmit that I am satisfied as it is based on a real problem and it achieves the objective statement: create a non-lineal objective function and one non trivial constraint. I have achieved that and I created more than one non trivial constraint.